### PR TITLE
DAOS-3884 API: Add API for conditional update feature.

### PR DESCRIPTION
**Pull request by mchaarawi**
Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

File: src/include/daos_obj.h [hunk 1]
```diff
@@ -105,6 +105,32 @@ enum {
 	DAOS_OF_MASK		= ((1 << OID_FMT_FEAT_BITS) - 1),
 };
 
+/** Number of bits reserved in IO flags bitmap for conditional checks.  */
+#define IO_FLAGS_COND_BITS	8
+
+enum {
+	/** Flag on Object */
+	DAOS_OP_OBJ		= (1 << 0),
```
Comment (jolivier23) [inline src/include/daos_obj.h]: Are there scenarios where we would want something like a conditional update on the akey but a conditional insert on the dkey?   There doesn't seem to be any way to specify such if you specify the flags in this way

File: src/include/daos_obj.h [hunk 2]
```diff
@@ -586,9 +621,9 @@ daos_obj_fetch(daos_handle_t oh, daos_handle_t th, daos_key_t *dkey,
  *					aggregated. Update result is undefined.
  */
 int
-daos_obj_update(daos_handle_t oh, daos_handle_t th, daos_key_t *dkey,
-		unsigned int nr, daos_iod_t *iods, d_sg_list_t *sgls,
-		daos_event_t *ev);
+daos_obj_update(daos_handle_t oh, daos_handle_t th, uint32_t flags,
```
Comment (jolivier23) [inline src/include/daos_obj.h]: 64 bits will be placed on the stack for alignment here anyway so maybe uint64_t for forward compatibility?

File: src/include/daos_obj.h (hunk 1 above)
Comment (mchaarawi) [reply src/include/daos_obj.h]: conditional insert on dkey, means fail if dkey exists. so the akey check doesn't make sense.

i guess conditional update on dkey and conditional insert on akey, might make sense, although not sure if there is a real use case for it. maybe we have to go with my other proposal (from mailing list)

File: src/include/daos_obj.h (hunk 2 above)
Comment (mchaarawi) [reply src/include/daos_obj.h]: agreed, will update

**Pull request merged**